### PR TITLE
Support scalar table functions (double unnest of ARRAY<ROW>)

### PR DESCRIPTION
**Pull request by kokosing**
Support scalar table functions (double unnest of ARRAY<ROW>)

### CODE REVIEW

File: presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4 [hunk 1]
```diff
@@ -224,10 +224,10 @@ columnAliases
     ;
 
 relationPrimary
-    : qualifiedName                                                   #tableName
-    | '(' query ')'                                                   #subqueryRelation
-    | UNNEST '(' expression (',' expression)* ')' (WITH ORDINALITY)?  #unnest
-    | '(' relation ')'                                                #parenthesizedRelation
+    : qualifiedName                                                             #tableName
+    | '(' query ')'                                                             #subqueryRelation
+    | (TABLE | UNNEST) '(' expression (',' expression)* ')' (WITH ORDINALITY)?  #unnest
```
Comment (kokosing) [inline presto-parser/src/main/antlr4/com/facebook/presto/sql/parser/SqlBase.g4]: But does it conflict with anything? I thought it could be a nice extension to standard as it is already implemented for `UNNEST`.

Is there any problem of having this, or is just non standard compliant?

**Comment by kokosing** [comment]
Rebased and comments addressed.

**Comment by kokosing** [comment]
Rebased and resolved conflicts. Ping.

**Comment by kokosing** [comment]
@martint Can you please take a look. It has been waiting for a while, and it costs a rebasing effort.
If it is not on your priority list for the near future, then I would prefer to close it to avoid pointless rebasing.

**Comment by martint** [comment]
My understanding from the spec is that `TABLE (...)` is exactly like UNNEST, except that it only allows a routine invocation as its single argument:

```
6) If TP simply contains a <table function derived table> TFDT, then:
   a) The <collection value expression> immediately contained in TFDT shall be a <routine invocation>.
   b) Let CN be the <correlation name> simply contained in TP.
   c) Let CVE be the <collection value expression> simply contained in TP.
   d) Case:
      i) If TP specifies a <derived column list> DCL, then let TFDCL be 
              ( DCL )
      ii) Otherwise, let TFDCL be a zero-length string.
   e) TP is equivalent to the <table primary>
           UNNEST ( CVE ) AS CN TFDCL
```

**Comment by kokosing** [comment]
Currently TABLE accepts a list of expressions, would be ok to you to limit that just to a single expression? IMO limiting that further to just function call to match the spec is a kind of an overkill, with expression it is much more expressive with no costs. 

IMO having just a function is not much simpler as function have a list of arguments (expression) which still can be very complicated.

**Comment by martint** [comment]
Yeah, I wouldn't limit to a function, but I would probably keep it to a single expression (at least for now) since multiple expressions would require different semantics.

File: presto-main/src/main/java/com/facebook/presto/operator/ArrayOfRowsUnnester.java [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+
+        this.arrayBlock = arrayBlock;
+        this.positionCount = arrayBlock == null ? 0 : arrayBlock.getPositionCount();
+    }
+
+    protected void appendTo(PageBuilder pageBuilder, int outputChannelOffset)
+    {
+        Block rowBlock = arrayBlock.getObject(position++, Block.class);
+        for (int i = 0; i < elementType.getFields().size(); ++i) {
+            BlockBuilder bb = pageBuilder.getBlockBuilder(outputChannelOffset + i);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/ArrayOfRowsUnnester.java]: don't use abbreviation

File: presto-main/src/main/java/com/facebook/presto/operator/ArrayOfRowsUnnester.java [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+        requireNonNull(arrayType, "arrayType is null").getElementType();
+        Preconditions.checkArgument(arrayType.getElementType() instanceof RowType, "expected element type to be RowType");
+        elementType = (RowType) arrayType.getElementType();
+        this.channelCount = elementType.getFields().size();
+
+        this.arrayBlock = arrayBlock;
+        this.positionCount = arrayBlock == null ? 0 : arrayBlock.getPositionCount();
+    }
+
+    protected void appendTo(PageBuilder pageBuilder, int outputChannelOffset)
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/operator/ArrayOfRowsUnnester.java]: why protected and not private?

File: presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java [hunk 4]
```diff
@@ -601,16 +578,21 @@ protected Scope visitUnnest(Unnest node, Scope scope)
             ImmutableList.Builder<Field> outputFields = ImmutableList.builder();
             for (Expression expression : node.getExpressions()) {
                 ExpressionAnalysis expressionAnalysis = analyzeExpression(expression, scope);
-                Type expressionType = expressionAnalysis.getType(expression);
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/analyzer/StatementAnalyzer.java]: this is some kind of refactor/rebase I think. Shouldn't be part of the commit

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 5]
```diff
@@ -431,11 +431,7 @@ protected RelationPlan visitQuerySpecification(QuerySpecification node, Void con
     protected RelationPlan visitValues(Values node, Void context)
     {
         Scope scope = analysis.getScope(node);
-        ImmutableList.Builder<Symbol> outputSymbolsBuilder = ImmutableList.builder();
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: In the future only `RelationType` (not the `Scope` itself) needs to be stored for `Node` in `Analysis`.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 6]
```diff
@@ -390,20 +390,34 @@ private RelationPlan planCrossJoinUnnest(RelationPlan leftPlan, Join joinNode, U
...
-            if (type instanceof ArrayType) {
-                unnestSymbols.put(inputSymbol, ImmutableList.of(unnestedSymbolsIterator.next()));
-            }
-            else if (type instanceof MapType) {
-                unnestSymbols.put(inputSymbol, ImmutableList.of(unnestedSymbolsIterator.next(), unnestedSymbolsIterator.next()));
+            if (node.isUnnestTable()) {
+                if (((ArrayType) type).getElementType() instanceof RowType) {
+                    RowType elementType = (RowType) ((ArrayType) type).getElementType();
+                    List<Symbol> arrayUnnestedSymbols = elementType.getFields().stream()
+                            .map(f -> unnestedSymbolsIterator.next())
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: I would use simple loop here instead of mixing streams with iterators.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 7]
```diff
@@ -503,6 +494,13 @@ else if (type instanceof MapType) {
         return new RelationPlan(unnestNode, scope, unnestedSymbols);
     }
 
+    private List<Symbol> getOutputSymbols(Scope scope)
+    {
+        return scope.getRelationType().getVisibleFields().stream()
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: This should actually use: `analysis.getOutputDescriptor(node);`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 8]
```diff
@@ -476,21 +490,35 @@ protected RelationPlan visitUnnest(Unnest node, Void context)
...
-            if (type instanceof ArrayType) {
-                unnestSymbols.put(inputSymbol, ImmutableList.of(unnestedSymbolsIterator.next()));
-            }
-            else if (type instanceof MapType) {
-                unnestSymbols.put(inputSymbol, ImmutableList.of(unnestedSymbolsIterator.next(), unnestedSymbolsIterator.next()));
+            if (node.isUnnestTable()) {
+                if (((ArrayType) type).getElementType() instanceof RowType) {
+                    RowType elementType = (RowType) ((ArrayType) type).getElementType();
+                    List<Symbol> arrayUnnestedSymbols = elementType.getFields().stream()
+                            .map(f -> unnestedSymbolsIterator.next())
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: I would use simple loop here instead of mixing streams with iterators.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 9]
```diff
@@ -476,21 +490,35 @@ protected RelationPlan visitUnnest(Unnest node, Void context)
             values.add(LiteralInterpreter.toExpression(constantValue, type));
             Symbol inputSymbol = symbolAllocator.newSymbol(expression, type);
             argumentSymbols.add(inputSymbol);
-            if (type instanceof ArrayType) {
-                unnestSymbols.put(inputSymbol, ImmutableList.of(unnestedSymbolsIterator.next()));
-            }
-            else if (type instanceof MapType) {
-                unnestSymbols.put(inputSymbol, ImmutableList.of(unnestedSymbolsIterator.next(), unnestedSymbolsIterator.next()));
+            if (node.isUnnestTable()) {
```
Comment (sopel39) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: section building `unnestSymbols` from `visitUnnest` and `planCrossJoinUnnest` can be extracted into separate method

**Comment by martint** [comment]
> TP is equivalent to the <table primary>
>           UNNEST ( CVE ) AS CN TFDCL

I spent some more time reading the spec, and it turns out that TABLE is strictly syntactic sugar for UNNEST. When the type of the elements passed to unnest is a row type, the operation is supposed to flatten the fields of the rows into the output. I filed a separate issue for this: https://github.com/prestodb/presto/issues/8151
